### PR TITLE
Create pluggable data retriever pattern

### LINKED ISSUE #116: Abstract data collection
Abstract collection of usage data into `Retriever` plugins. This will add the ability to collect data from more sources (e.g. CloudTrail) or in different ways (e.g. IAM Access Advisor data via AWS Organizations).

### PR DESCRIPTION
**Pull request by patricksanders**
This PR adds support for `AardvarkRetriever` plugins to get usage data for AWS principals. It also refactors the existing updater logic into an AccessAdvisorRetriever plugin.

Closes #116

### CODE REVIEW

File: aardvark/retrievers/runner.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            log.error(
+                "Account names passed but SWAG not configured or unavailable: {}".format(
+                    e
+                )
+            )
+
+        return all_accounts
+
+    async def _queue_all_accounts(self):
+        for account in await self._get_swag_accounts():
```
Comment (saillinux) [inline aardvark/retrievers/runner.py:111]: is this need to be await for _get_swag_accounts call? I don't see any asynchronous call in the function unless SWAG back-end does it.

File: aardvark/retrievers/runner.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            await self.account_queue.put(account["id"])
+
+    async def _queue_accounts(self, account_names: List[str]):
+        accounts = copy(account_names)
+        for account in accounts:
+            if re.match(r"\d{12}", account):
+                accounts.remove(account)
+                await self.account_queue.put(account)
+
+        all_accounts = await self._get_swag_accounts()
```
Comment (saillinux) [inline aardvark/retrievers/runner.py:121]: same here

File: aardvark/retrievers/runner.py (hunk 1 above)
Comment (patricksanders) [reply aardvark/retrievers/runner.py:118]: Wrapped the SWAG calls with `sync_to_async`

File: aardvark/manage.py [hunk 3]
```diff
@@ -35,60 +30,6 @@
 DEFAULT_NUM_THREADS = 5  # testing shows problems with more than 6 threads
```
Comment (Dreadwall) [inline aardvark/manage.py:30]: Nitpick: update comment

File: aardvark/retrievers/runner.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+        Retriever plugins are executed in the order in which they are registered. Each retriever
+        is passed the result from the previous one, starting with a dict containing an `arn` element.
+
+        Note: The data from all previous retriever plugins is mutable by subsequent ones.
+        """
+        data = {
+            "arn": arn,
+        }
+        # Iterate through retrievers, passing the results from the previous to the next.
```
Comment (Dreadwall) [inline aardvark/retrievers/runner.py:68]: So retrievers are acting as compounding gathering and enrichment pipelines?

Comment (Dreadwall) [reply aardvark/retrievers/runner.py:68]: Will this play nicely with the Aardvark Digest pattern we are looking into?

**Review by Dreadwall — APPROVED** [review]

File: aardvark/retrievers/runner.py (hunk 4 above)
Comment (patricksanders) [reply aardvark/retrievers/runner.py:68]: Yeah, that's a good way to describe them. Subsequent retrievers can do whatever they want to the data dict that's passed in to them (for better or worse). So my thought is for the action usage retriever to enrich the Access Advisor data in a similar way to how action-level data is returned from AA for S3 (i.e. in the nested `TrackedActionsLastAccessed` list for each service).

**Pull request merged**